make stringify public

morganchen12 opened the pull request:
An attempt at #258.

A few things to note:
- I removed the `stringify` function for `SequenceType` because `SequenceType` doesn't make guarantees about what its generator does, so `stringify`ing a `SequenceType` that maintained some sort of internal state could have unexpected side effects. `SequenceType`s can still be stringified via the top-level function or by individuals manually conforming their sequences to `Stringifiable`.

- I changed the behavior slightly for printing arrays with strings in them. Previously Nimble would wrap the strings in an array in double quotes. Now they don't. Strings aren't wrapped in double quotes anywhere else, and inside of the array they seemed redundant. But maybe they're not, and I messed up--please let me know!

- `Stringifiable` is a really vague protocol name; maybe we should imitate the standard library and call our protocol `TestOutputStringConvertible`. Or something else.


As always, feedback is welcome!

jeffh commented:
Hey @morganchen12,

Just getting back in the groove from jet lag.

- Stringifiers for SequenceTypes were suppose to keep the format looking similar to Arrays. I haven't recently verified the output, but generators don't usually return useful truing values by default.
- The quoting of strings was an artifact from the old dates. Ideally, Nimble could print strings that could be copy-pasted as code. But that's probably nice for some theoretical future. It's fine that you removed it.
- Agreed. `TestOutputStringConvertible` sounds good to me! :+1:

morganchen12 commented:
- I changed the behavior of printing all numbers to drop trailing zeros behind the decimal if they were unnecessary and round to four points of precision only if there were four or more digits behind the decimal. In hindsight I'm not sure why I did this. But it does mean that numbers are now represented consistently across platforms, and I got to delete a lot of (like 3) `#if _runtime(_ObjC)` clauses in our tests!

- Adding the quotes back in would be very easy since Swift's `string.debugDescription` returns itself padded in quotes already. 

- Renamed `Stringifiable` to `TestDebugStringConvertible`.

jeffh commented on Tests/Nimble/objc/ObjcStringersTest.m:
```diff
@@ -0,0 +1,24 @@
+    NSArray *array = @[@1, @2, @3];
+    NSString *result = NMBStringify(array);
+    
+    XCTAssert([result isEqualToString:@"(1, 2, 3)"], @"got <%@>, expected <(1, 2, 3)>", result);
```
I don't see an issue for using Nimble expectations here. Please use it instead of XCTest to be more consistent

morganchen12 replied:
😅 whoops

jeffh commented:
Thanks @morganchen12!
